app/com/linkedin/drelephant/exceptions/TonYExceptionFingerprinting.java

```diff
@@ -361,4 +368,56 @@ private boolean isExceptionLogSimilar(String exceptionStackTrace) {
     }
     return false;
   }
+
+
+  /**
+   * @param stackTrace String containing stackTrace of exception seperated by new line
+   * @param limit maximum number of lines required in stackTrace
+   * @return stackTrace with given number of lines or less
+   */
+  private String truncateStackTrace(String stackTrace, int limit) {
+    List<String> stackTraceSplitList = Arrays.asList(stackTrace.split("\n"));
+    return String.join("\n", stackTraceSplitList.subList(0, Math.min(limit, stackTraceSplitList.size())));
+  }
+
+  /**
+   *
+   * @return Return the classification for the exception .
+   */
+  public String classifyException() {
+    if (EXCEPTION_CATEGORIZATION_CONFIGURATION == null || this._exceptionInfoList == null
+        || this._exceptionInfoList.size() == 0) {
+      logger.error(" Necessary information is not available for classifying exceptions ");
+      return ERROR_CLASSIFICATION;
+    }
+    List<ExceptionCategorizationData> exceptionCategorizationData =
+        EXCEPTION_CATEGORIZATION_CONFIGURATION.getValue().getExceptionCategorizationData().get(APPLICATION_TYPE);
+    if (exceptionCategorizationData == null || exceptionCategorizationData.size() == 0) {
+      logger.error(" No exception categorization rules for tony application ");
+      return ERROR_CLASSIFICATION;
+    }
+    String className = UNKNOWN_CLASSIFICATION;
```

Comment:
Why `UNKNOWN_CLASSIFICATION` classification is User induced error(`USER_ERROR/UNKNOWN`)?
Will using `UNKNOWN_CLASSIFICATION = "UNKNOWN_ERROR"`  won't be suffice?

Reply:
Because , we know the universe  of TONY_INFRA_ERROR , TENSOR_FLOW_ERROR but not USER ERROR . And its being found out that , the error which are not been able to classified are mostly USER ERROR . . Thats why the default classification is USER_ERROR/UNKNOWN. 

Please also see the doc in which it has been shown clearly

Reply:
NIT: I am not able to access the Doc from my personal account. If possible, make the doc visible to the public(after removing all company-related data) as it is part of PR description.

Reply:
I don't think so , doc can be made public it has LinkedIn specific information (its similar to my earlier PR , doc is attached for more info) .  I can create public doc from next time , if it required (this will be seperate effort)

Reply:
Kindly don't mind, but I am able to properly concur with the reason 
`We know the universe of TONY_INFRA_ERROR, TENSOR_FLOW_ERROR` if this means that the Categorization module has all exceptions covered for TONY and INFRA error.

Also, I tried to scrape through the doc and came across this statement
`Around 17% of users(assuming here that the author wants to say error) are unclassified, which requires more effort to further classify them. Note, these errors may also fall to tony and tensor flow error.`
The suggestion is not to make `UNKNOWN_CLASSIFICATION` a subCategory of any category of exceptions( TONY_INFRA_ERROR, TENSOR_FLOW_ERROR, USER_ERROR) and just make it as an individual category for errors which can't be classified.

Reply:
Also, I tried to scrape through the doc and came across this statement
Around 17% of users(assuming here that the author wants to say error) are unclassified, which requires more effort to further classify them. Note, these errors may also fall to tony and tensor flow error.


Response : Nope these error will not fall into tony and tensor flow error , as we covered all the universe of both of them . These 17% will only fall into USER_ERROR category . 

As I already told UNKNOWN_CLASSIFICATION can not be general classification , since as I earlier told , since we know what are the exceptions should fall in tensor flow and tony . 
If out of 100 UNKNOWN , 99 falls into user category then it make sense to default them to USER/UNKNOWN (infect in analysis all 100  are part of USER) . 

Just making all exception UNKNOWN , will have serious impact to user , since he/she will think all unknown might falls  in Tony & Tensor flow (whereas its opposite in current scenarios) and will create unnecessary jira .

Reply:
**pralabhkumar: `Nope these error will not fall into tony and tensor flow error, as we covered all the universe of both of them`**

Not sure how `UNIVERSE` of TENSOW_FLOW_ERROR and especially INFRA_ERROR can be covered by mere (10-15) types of error.

One example: Suppose `NoSuchMethodFound` exception (took this as an example as it is now present in any rule so UNKNOWN) has occurred, then can it be certainly said that it is a USER_ERROR as this can be caused by both User and Platform(TonY).

Reply:
Yeah, making UNKOWN_ERROR as a separate category may lead to queries from users to the Platform team. But one upside is that these queries can act as a penalty for the classification model and will increase the knowledge base of it, in turn improving the classification.
Also, any false positive(TENSORFLOW/TONY exception classified as UNKNOWN/USER_ERROR) can lead to loss of user's time and productivity which is more impactful  (assuming user's time is more valuable) than some unnecessary queries.

Reply:
This PR is the implementation of the doc , and based on the cases which we have seen . Classification is always the evolution process . It cannot be 100% accurate in the first time. Thats the reason rules are configurable . They will/should change in future .  As again , this PR , is just the implementation of the results shared with Tony Team , its no way saying that classification is absolutely correct.   Every classification system , is build on feedback and lots of iteration . But it starts sample data set get manually annotated (Results of which are in doc) . 

Suppose NoSuchMethodFound exception (UNKNOWN) has occurred, then can it be certainly said that it is a USER_ERROR as this can be caused by both User and Platform(TonY) 

In case of TonY , we have not seen this exception in sample data set , and that's why the reason it is not in rules. Now for you assumption it can USER_ERROR or TonY error is correct . But as per the sample set and discussion in doc , chances of being it USER_ERROR is more than TonY error and thats why default categorization is USER_ERROR/UNKNOWN.   Later on , if we see lot of USER_ERROR/UNKNOWN is actually not correct ,then rules should be changed and thats what is iterative process.

Reply:
In the doc , requirements are clearly specified by TonY team (which we can assume source of truth) . 
--Requirement 
1) tony errors: some java exceptions most likely.
2) tensorflow errors: tf crashed. normally c++ memory errors
3) user errors: anything other than 1 & 2.

which is what is implemented as part of this PR .

Reply:
**Also, any false positive(TENSORFLOW/TONY exception classified as UNKNOWN/USER_ERROR) can lead to loss of user's time and productivity which is more impactful (assuming user's time is more valuable) than some unnecessary queries.**

I think , the point which is being missed here is probability of any unknown error to be USER error is very high then of being Tensorflow / tony error ,which will lead to far more false positive if its classified as UNKNOWN (as major chunk of unknown will be user error) . 

**But one upside is that these queries can act as a penalty for the classification model and will increase the knowledge base of it, in turn improving the classification**

That depends on the post analysis + feedback , which should be done for any category , not just one classification .